### ISSUE TITLE
jsx cast cannot be parsed!

### ISSUE THREAD

**Issue by 929796111** [evanw/esbuild#1335]
hi, I have a jsx code segment in my .vue file  as followed:

```jsx
function colFormItem(h, currentItem, index, list) {
......
return (<div>222<button>1111</button>333</div>)
}
```

it works ,but wrong!
when i changed to
```jsx
<div>222</div>
```
it works fine;
I think jsx cast cause this error,
my stack is electron-vue+element-ui, my webpack configuration such as this:
```js
{
        test: /\.js$/,
        loader: 'esbuild-loader',
        options: {
          loader: 'jsx',
          target: 'es2015',
          jsxFactory: 'h'
        }
      }
```
how can i fix this problem? thanks!

**Comment by evanw** [comment]
> but wrong!

What does "but wrong" mean? What is wrong?

**Comment by 929796111** [comment]
function colFormItem(h, currentItem, index, list) {
......
return (<div>222<button>1111</button>333</div>)
}
only output '<div>222</div'> on html page ,   '<button>1111</button>333 'not parsed and losed;
it seemed like that all cast tags and followed cast tags are not parsed and dropped.

**Comment by evanw** [comment]
It looks like esbuild can handle this fine:

```
$ cat in.js | ./esbuild --loader=jsx --target=es2015 --jsx-factory=h
function colFormItem(h, currentItem, index, list) {
  return /* @__PURE__ */ h("div", null, "222", /* @__PURE__ */ h("button", null, "1111"), "333");
}
```

The `1111` and `333` are both present so the output appears to be correct. If you are running into a problem here then I'm guessing the problem is with whatever tools or libraries you are using after esbuild runs, not with esbuild itself.

**Comment by 929796111** [comment]
Thanks for your reminding me，I tried like this:
return(
      h("el-col", {span:config.span, class:className, nativeOnClick:event => { activeItem(currentItem); event.stopPropagation() }},
          [h("render", {key:config.renderKey, props:{conf:currentItem}, onInput: event => {this.$set(config, 'defaultValue', event)}}, child),components.itemBtns.apply(this, arguments)]
    )
    )
it works fine, but this is out of jsx, of some reason, i hope these html segments can be easy read and understand. 
the following code can worked fine with babel-loader, but worked wrong with esbuilder:
`return (
      <el-col span={config.span}>
        <button ......>  
      </el-col>
    )`
I thinks this maybe jsx-processor not process inner tags. should recursive jsx-code with jsx process function.

**Comment by evanw** [comment]
It looks like maybe you are using a library that requires a different JSX transform than the one built into esbuild? Your example code puts nested elements in an array as a third argument to `h` while esbuild follows the React-style JSX transform and puts nested elements as additional arguments following the second argument:

```js
// Original code
<div>222<button>1111</button>333</div>

// Output from esbuild
h("div", null, "222", h("button", null, "1111"), "333");

// Maybe the output you need?
h("div", null, ["222", h("button", null, "1111"), "333"]);
```

If this is the problem, you could potentially shim it with a function like this:

```js
export const h_shim = (tag, props, ...children) => h(tag, props, children)
```

That could then be injected using the [inject](https://esbuild.github.io/api/#inject) feature.

But if the problem is something else, then it many not be possible to adapt the JSX transform that esbuild uses to the JSX transform that you need. In that case you'd still have to run whatever JSX transform you're currently using.

**Comment by evanw** [comment]
I checked how Babel handles this code and esbuild appears to be transforming JSX exactly the same as Babel.

Here's the output from esbuild:

```js
function colFormItem(h, currentItem, index, list) {
  return /* @__PURE__ */ React.createElement("div", null, "222", /* @__PURE__ */ React.createElement("button", null, "1111"), "333");
}
```

Here's the output from Babel:

```js
function colFormItem(h, currentItem, index, list) {
  return /*#__PURE__*/React.createElement("div", null, "222", /*#__PURE__*/React.createElement("button", null, "1111"), "333");
}
```

You can see this for yourself on the Babel REPL: https://babeljs.io/en/repl#?browsers=defaults%2C%20not%20ie%2011%2C%20not%20ie_mob%2011&build=&builtIns=false&corejs=3.6&spec=false&loose=false&code_lz=GYVwdgxgLglg9mABBOAbAYnATgWwJJQCmOAFABYA0yIWWhYUBxVMYAJoQB5WowDOUAJSIA3gCg6UGkhIAeNjABuAPgBM62QCMQUKAmUBGIwdkB6bbv0BmG2YUrBYgL5igA&debug=false&forceAllTransforms=false&shippedProposals=false&circleciRepo=&evaluate=false&fileSize=false&timeTravel=false&sourceType=module&lineWrap=true&presets=env%2Creact%2Cstage-2&prettier=false&targets=&version=7.14.4&externalPlugins=

I'm closing this issue as "working as intended." If you want a custom JSX transform that differs from how esbuild works, then you can either use a shim function as described above or an [on-load plugin](https://esbuild.github.io/plugins/#load-callbacks) to run your custom JSX transform before esbuild bundles the code.

**Comment by 929796111** [comment]
thanks，now It's very nearly root cause，yes，this is I need：
’h("div", null, ["222", h("button", null, "1111"), "333"]);‘
but render to h("div", null, "222", h("button", null, "1111"), "333");
 I used Vue, It's render function like 
h("div", null, ["222", h("button", null, "1111"), "333"]);
the third parameter must be array, contains its childen.

I found some words in esbuild api document:
This loader is enabled by default for .jsx and .tsx files. Note that JSX syntax is not enabled in .js files by default. If you would like to enable that, you will need to configure it:

package main

import "github.com/evanw/esbuild/pkg/api"
import "os"

func main() {
  result := api.Build(api.BuildOptions{
    EntryPoints: []string{"app.js"},
    Bundle:      true,
    Loader: map[string]api.Loader{
      ".js": api.LoaderJSX,
    },
    Write: true,
  })

  if len(result.Errors) > 0 {
    os.Exit(1)
  }
}
it's just my since, I used webpack, include esbuild-loader for process .js file(not .jsx file, just contains some jsx code in .js file)
my webpack config like this :

......
let rendererConfig = {
  devtool: '#cheap-module-eval-source-map',
  entry: IsWeb ? { web: path.join(__dirname, '../src/renderer/main.js')} : { renderer: resolve('src/renderer/main.js') },
  externals: IsWeb ? [] : [...Object.keys(dependencies || {}).filter(d => !whiteListedModules.includes(d))],
  module: {
    rules: [
      {
        test: /\.html$/,
        use: 'vue-html-loader'
      },
      {
        test: /\.js$/,
        loader: 'esbuild-loader',
        options: {
          loader: 'jsx',
          target: 'es2015',
          jsxFactory: 'h'
        }
      },
......
is there any way inject h_shim js code to esbuild  loader in this config file, not in cmd console?
